src/Nethermind/Nethermind.Serialization.Rlp/ReceiptArrayStorageDecoder.cs

```diff
@@ -0,0 +1,156 @@
+        }
+        return bufferLength;
+    }
+
+    private int GetContentLength(TxReceipt[] items, RlpBehaviors rlpBehaviors)
+    {
+        if (_useCompactEncoding && (rlpBehaviors & RlpBehaviors.Storage) != 0)
+        {
+            int totalLength = 0;
+            for (int i = 0; i < items.Length; i++)
+            {
+                totalLength += CompactReceiptStorageDecoder.GetLength(items[i], rlpBehaviors);
+            }
+
+            return totalLength;
+        }
+        else
+        {
+            int totalLength = 0;
+            for (int i = 0; i < items.Length; i++)
+            {
+                totalLength += StorageDecoder.GetLength(items[i], rlpBehaviors);
+            }
+
+            return totalLength;
+        }
+    }
+
+    public TxReceipt[] Decode(RlpStream rlpStream, RlpBehaviors rlpBehaviors = RlpBehaviors.None)
+    {
+        if (rlpStream.PeekByte() == CompactEncoding)
+        {
+            rlpStream.ReadByte();
+            return CompactReceiptStorageDecoder.DecodeArray(rlpStream, RlpBehaviors.Storage);
+        }
+        else
+        {
+            return StorageDecoder.DecodeArray(rlpStream, RlpBehaviors.Storage);
+        }
+    }
+
+    public void Encode(RlpStream stream, TxReceipt[] items, RlpBehaviors rlpBehaviors = RlpBehaviors.None)
+    {
+        if (items is null || items.Length == 0)
+        {
+            stream.EncodeNullObject();
+            return;
+        }
+
+        if (_useCompactEncoding && (rlpBehaviors & RlpBehaviors.Storage) != 0)
+        {
+            int totalLength = GetContentLength(items, rlpBehaviors);
+            stream.WriteByte(CompactEncoding);
+            stream.StartSequence(totalLength - 1);
+
+            for (int i = 0; i < items.Length; i++)
+            {
+                CompactReceiptStorageDecoder.Encode(stream, items[i], rlpBehaviors);
+            }
+        }
+        else
+        {
+            int totalLength = GetContentLength(items, rlpBehaviors);
+            stream.StartSequence(totalLength);
+
+            for (int i = 0; i < items.Length; i++)
+            {
+                StorageDecoder.Encode(stream, items[i], rlpBehaviors);
+            }
+        }
+    }
+
+    public TxReceipt[] Decode(in Span<byte> receiptsData)
+    {
+        if (receiptsData.Length == 0 || receiptsData[0] == Rlp.NullObjectByte)
+        {
+            return Array.Empty<TxReceipt>();
+        }
+
+        if (receiptsData.Length > 0 && receiptsData[0] == CompactEncoding)
+        {
+            var decoderContext = new Rlp.ValueDecoderContext(receiptsData.Slice(1));
+            return CompactReceiptStorageDecoder.DecodeArray(ref decoderContext, RlpBehaviors.Storage);
+        }
+        else
+        {
+            var decoderContext = new Rlp.ValueDecoderContext(receiptsData);
+            try
+            {
+                return StorageDecoder.DecodeArray(ref decoderContext, RlpBehaviors.Storage);
+            }
+            catch (RlpException)
```

Comment:
Is this a common or rare path? Can it be handled without going via exception handling?

Reply:
Throwing and catching exceptions are very expensive

Reply:
I don't exactly know, come from previous code. Seems to be it might store the receipt in network encoding. @LukaszRozmej probably know.

Reply:
It is a remnant of very old code, could be actually removed.